### ISSUE TITLE
Ability to disable hook reactions temporarily

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#470]
Previously: https://github.com/WordPoints/hooks-api/issues/11

This will be necessary when we remove the old Points Hooks administration screen (#389), because of the inactive points hooks feature.

**Comment by JDGrimes** [comment]
Note that in theory hook reactions for the points reactor can already be temporarily disabled by setting the points value to `0`, since this is valid, I think. However, this doesn't actually disable, them, it only prevents them from firing—reverse fires can continue.

I suppose that this is something that we'll need to consider, should disabling disable reverse fires as well?

**Comment by JDGrimes** [comment]
I don't think that we can just use the blocker extension here, because we already use it for other things. When a reaction is disabled, it should not interfere with any other settings, which would be the case if we used the blocker extension, and there were already some settings for it.

What we could do however, is just sub-class the blocker extension, and give it a different slug. Actually, if we allowed the extensions to be constructed with their slugs instead of defining them internally, we could just register the existing class as a new extension with a different slug.

**Comment by JDGrimes** [comment]
Here is what it would look like with the button on the title bar for the reaction:

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-15_15 53 33](https://cloud.githubusercontent.com/assets/4005415/22994988/eec711a2-f396-11e6-8338-ac55f58a7cd2.png)

This is the way that I've kind of always envisioned it, though I have also questioned whether we should put it "inside" the reaction instead. That would force the user to click the edit button to display the reaction form, then scroll to the enable/disable button, if necessary (assuming it was at the bottom), just to enable/disable the reaction. But forcing the user to actually look over the reaction settings might be a good idea. I'd guess that in many user's workflows relating to this, they'd want to open the settings just to double-check things anyway.

Another argument in favor of that is that this is an operation that actually "modifies" the reaction, that is, it touches the database. Or, I guess that it would if we don't require the user to also hit the save button; I was thinking if it was on the title bar, saving would be automatic. I guess if it wasn't, then there would be no need to have it be a button then, just let it be a checkbox or something like that.

But the point I was starting out to make there was that the save and delete buttons are both only shown in the action bar at the bottom of the form, and the title bar only shows either the edit or close or cancel button. In other words, the title bar only contains controls for the form, whereas the action bar at the bottom also contains controls for the reaction itself. 

So all things considered, although I've always thought of this as a enable/disable button, we might want to explore just making it a checkbox at the top/bottom of the reaction form instead.

**Comment by JDGrimes** [comment]
Here is what it could look like at the top of the reaction form:

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-15_16 13 54](https://cloud.githubusercontent.com/assets/4005415/22995668/c94f4eaa-f399-11e6-8f0b-6cdd380c40cd.png)

I dislike this, and think it makes things too cluttered. That was the beauty of a button. Let's try it at the bottom though, and maybe with its own "heading" like the conditions and rate limits extensions have.

**Comment by JDGrimes** [comment]
Here is what it looks like at the bottom without its own heading:

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-15_16 17 36](https://cloud.githubusercontent.com/assets/4005415/22995803/52e5100a-f39a-11e6-8f7f-ebfde8a8af0c.png)

And with it:

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-15_16 19 51](https://cloud.githubusercontent.com/assets/4005415/22995877/9d621f56-f39a-11e6-9a21-bdb6a288dc4f.png)

I think I like it with the heading better, although I'm not sure about the text. Neither just "Disable" nor "Enable/disable" seem quite right. Maybe something better will come to us.

**Comment by JDGrimes** [comment]
Just "Disable" looks funny though, when the label of the checkbox is "Enabled". 

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-15_16 30 36](https://cloud.githubusercontent.com/assets/4005415/22996262/2a6918cc-f39c-11e6-86f1-a010dd29b587.png)

**Comment by JDGrimes** [comment]
We could change the checkbox to be "Disable" instead of "Enabled" though. Actually, that makes more sense, since the extension settings will disable the reaction when the setting is set, not when it is not. 

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-15_16 41 01](https://cloud.githubusercontent.com/assets/4005415/22996660/8edc6146-f39d-11e6-8e61-3935af867442.png)

I like that better.

The only downside to this entire approach is that we are adding more vertical content to the reaction settings form, which will cause folks to need to scroll more, especially on small screens. But let's experiment with doing it like this for now, and if we need to revisit it later, we can do it differently then.

**Comment by JDGrimes** [comment]
I would like to keep this as the last thing in the form, just above the bottom action bar. Originally I was just going to hard-code it below the general `.fields` container. However, I decided that it would be better to make it a separate view and controller that is dynamically appended to the form, like we do with the other extensions. The thing that kind of prompted me to go ahead and do that was the thought that otherwise the extension could be deregistered internally (by someone who didn't want to use it for some reason), but it wouldn't affect the UI. This would also make it harder to extend/replace the extension & its UI. I also realized that not having the checkbox inside of the `.fields` container would cause it to not automatically be included with the other fields that the reaction view grabs when the Save button is clicked. 

Anyway, the trouble is that the position of the settings in the form is just determined by the order in which the extensions are registered, now that it isn't hard-coded in. But as I said, I'd like to keep this extension's settings at the bottom, regardless of what the other extensions are doing. There's currently no way to do that though.

My only idea was to append the extension's form last by having the function hooked to `render:fields` actually just hook another function to `render:fields`. That way that function would be hooked up last, and so would be the last to fire, and thus the extension's form would be the last to be appended. But it seems that callbacks that are registered after the fire of the Backbone event has already begun are not fired at the end.

I guess I could hook the hooker-upper function to `render` or `render:settings` instead, which fire earlier. It would really be a hack though, and eventually we'll need to implement some kind of a priority API here.

**Comment by JDGrimes** [comment]
Here is what a disabled reaction's title bar looks like:

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-16_15 00 19](https://cloud.githubusercontent.com/assets/4005415/23038622/b4872b90-f458-11e6-95cd-cd25b09fb3d8.png)

And of course, it can still be edited:

![wordpoints_ _points_types_ _wordpress_develop_ _wordpress_-_2017-02-16_15 02 07](https://cloud.githubusercontent.com/assets/4005415/23038689/f01af9c0-f458-11e6-8150-8df7b525910b.png)

**Comment by JDGrimes** [comment]
It seems that checkboxes pose an issue for us that we haven't dealt with yet in the reaction form. When the checkbox has been checked and the reaction saved, unchecking the checkbox will not delete the value when the reaction is saved again. The unchecked checkbox will basically be ignored when the values of all of the form fields are grabbed, and so the existing value will end up getting merged into the request parameters sent via Ajax.

I guess that really we should update the method that we use to get the form data so that it will handle checkboxes as expected.

**Comment by JDGrimes** [comment]
We missed one thing here, and that is that the disabled reactions should not be displayed in the HTGP shortcode. They currently are:

![wordpoints_how_to_get_points _ _wordpress_develop_-_2017-03-07_08 40 13](https://cloud.githubusercontent.com/assets/4005415/23658812/df8ae7e2-0311-11e7-879a-e77a0ef53391.png)

Let's fix that.
